pkg/controller/projects/deploykeys/controller.go

```diff
@@ -0,0 +1,265 @@
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+const (
+	ERR_NOT_DEPLOY_KEY = "managed resource is not a Gitlab deploy key custom resource"
+	ERR_NOT_FOUDN      = "404 project deploy key not found"
+	ERR_GET_FAIL       = "cannot get Gitlab deploy key"
+	ERR_CREATE_FAIL    = "cannot create Gitlab deploy key"
+	ERR_UPDATE_FAIL    = "cannot update Gitlab deploy key"
+	ERR_DELETE_FAIL    = "cannot delete Gitlab deploy key"
+	ERR_KEY_MISSING    = "missing key ref value"
+	ERR_ID_NOT_INT     = "external-name is not an int"
+)
+
+var (
+	deployKeyDoesNotExist = managed.ExternalObservation{}
+)
+
+type external struct {
+	kube   kubeClient.Client
+	client projects.DeployKeyClient
+}
+
+type connector struct {
+	kube              kubeClient.Client
+	newGitlabClientFn func(clientConfig clients.Config) projects.DeployKeyClient
+}
+
+func SetupDeployKey(manager kubeController.Manager, log logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.DeployKeyKind)
+
+	return kubeController.NewControllerManagedBy(manager).
+		Named(name).
+		For(&v1alpha1.DeployKey{}).
+		Complete(buildReconciler(manager, log, name))
+}
+
+func (deployKeyService *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+	customResource, ok := mgd.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return nil, errors.New(ERR_NOT_DEPLOY_KEY)
+	}
+
+	config, err := clients.GetConfig(ctx, deployKeyService.kube, customResource)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &external{kube: deployKeyService.kube, client: deployKeyService.newGitlabClientFn(*config)}, nil
+}
+
+func (deployKeyService *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	customResource, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return deployKeyDoesNotExist, errors.New(ERR_NOT_DEPLOY_KEY)
+	}
+
+	id := meta.GetExternalName(customResource)
+
+	if id == "" {
+		return deployKeyDoesNotExist, nil
+	}
+
+	optionList := &gitlab.ListProjectDeployKeysOptions{} // do i need those?
+	deployKeyArray, _, err := deployKeyService.client.ListProjectDeployKeys(*customResource.Spec.ForProvider.ProjectId, optionList)
+
+	if err != nil {
+		return deployKeyDoesNotExist,
+			errors.Wrap(resource.Ignore(isErrorProjectDeployKeyNotFound, err), ERR_GET_FAIL)
+	}
+
+	deployKey := findDeployKey(id, deployKeyArray)
+
+	if deployKey == nil {
+		return deployKeyDoesNotExist, nil
+	}
+
+	currentState := customResource.Spec.ForProvider.DeepCopy()
+
+	lateInitializeProjectDeployKey(&customResource.Spec.ForProvider, deployKey)
+
+	isLateInitialized := !compare.Equal(currentState, &customResource.Spec.ForProvider)
+
+	customResource.Status.AtProvider = v1alpha1.DeployKeyObservation{}
+
+	customResource.Status.SetConditions(xpv1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        true,
```

Comment:
Following this logic, the controller will never call `Update` since you always set `ResourceUpToDate` to `true`. You should implement a function `isUpToDate` that compares `cr` with `deployKey` and checks if any of the updatable fields has changed.